### PR TITLE
provider/aws: Add new "ANY" as valid HTTP method to API Gateway validator.

### PR DESCRIPTION
**Pull request by kwilczynski**
This commit adds a new HTTP method to a list of valid HTTP methods which
is now accepted by API Gateway.

Signed-off-by: Krzysztof Wilczynski <krzysztof.wilczynski@linux.com>

### CODE REVIEW

**Comment by kwilczynski** [comment]
The unit tests are passing:

```
$ make test TEST=./builtin/providers/aws TESTARGS='-v -run=TestSentenceJoin'
==> Checking that code complies with gofmt requirements...
==> Checking AWS provider for unchecked errors...
==> NOTE: at this time we only look for uncheck errors in the AWS package
go generate $(go list ./... | grep -v /terraform/vendor/)
2016/10/04 12:21:45 Generated command/internal_plugin_list.go
TF_ACC= go test ./builtin/providers/aws -v -run=TestSentenceJoin -timeout=30s -parallel=4
=== RUN   TestSentenceJoin
--- PASS: TestSentenceJoin (0.00s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/aws	0.017s

$ make test TEST=./builtin/providers/aws TESTARGS='-v -run=TestValidateHTTPMethod'
==> Checking that code complies with gofmt requirements...
==> Checking AWS provider for unchecked errors...
==> NOTE: at this time we only look for uncheck errors in the AWS package
go generate $(go list ./... | grep -v /terraform/vendor/)
2016/10/04 11:17:57 Generated command/internal_plugin_list.go
TF_ACC= go test ./builtin/providers/aws -v -run=TestValidateHTTPMethod -timeout=30s -parallel=4
=== RUN   TestValidateHTTPMethod
--- PASS: TestValidateHTTPMethod (0.00s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/aws	0.017s


$ make test TEST=./builtin/providers/aws TESTARGS='-v -run=TestResourceAWSS3BucketObjectAcl_validation'
==> Checking that code complies with gofmt requirements...
==> Checking AWS provider for unchecked errors...
==> NOTE: at this time we only look for uncheck errors in the AWS package
go generate $(go list ./... | grep -v /terraform/vendor/)
2016/10/04 11:22:36 Generated command/internal_plugin_list.go
TF_ACC= go test ./builtin/providers/aws -v -run=TestResourceAWSS3BucketObjectAcl_validation -timeout=30s -parallel=4
=== RUN   TestResourceAWSS3BucketObjectAcl_validation
--- PASS: TestResourceAWSS3BucketObjectAcl_validation (0.00s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/aws	0.016s
```

**Comment by kwilczynski** [comment]
The acceptance test is passing:

```
$ make testacc TEST=./builtin/providers/aws TESTARGS='-run=TestAccAWSAPIGateway'
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2016/10/04 11:38:28 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/aws -v -run=TestAccAWSAPIGateway -timeout 120m
=== RUN   TestAccAWSAPIGatewayAccount_importBasic
--- PASS: TestAccAWSAPIGatewayAccount_importBasic (27.72s)
=== RUN   TestAccAWSAPIGatewayApiKey_importBasic
--- PASS: TestAccAWSAPIGatewayApiKey_importBasic (54.24s)
=== RUN   TestAccAWSAPIGatewayAccount_basic
--- PASS: TestAccAWSAPIGatewayAccount_basic (156.46s)
=== RUN   TestAccAWSAPIGatewayApiKey_basic
--- PASS: TestAccAWSAPIGatewayApiKey_basic (53.32s)
=== RUN   TestAccAWSAPIGatewayAuthorizer_basic
--- PASS: TestAccAWSAPIGatewayAuthorizer_basic (76.79s)
=== RUN   TestAccAWSAPIGatewayBasePath_basic
--- PASS: TestAccAWSAPIGatewayBasePath_basic (56.99s)
=== RUN   TestAccAWSAPIGatewayClientCertificate_basic
--- PASS: TestAccAWSAPIGatewayClientCertificate_basic (44.33s)
=== RUN   TestAccAWSAPIGatewayClientCertificate_importBasic
--- PASS: TestAccAWSAPIGatewayClientCertificate_importBasic (30.27s)
=== RUN   TestAccAWSAPIGatewayDeployment_basic
--- PASS: TestAccAWSAPIGatewayDeployment_basic (51.96s)
=== RUN   TestAccAWSAPIGatewayDomainName_basic
--- PASS: TestAccAWSAPIGatewayDomainName_basic (33.28s)
=== RUN   TestAccAWSAPIGatewayIntegrationResponse_basic
--- PASS: TestAccAWSAPIGatewayIntegrationResponse_basic (80.33s)
=== RUN   TestAccAWSAPIGatewayIntegration_basic
--- PASS: TestAccAWSAPIGatewayIntegration_basic (62.62s)
=== RUN   TestAccAWSAPIGatewayMethodResponse_basic
--- PASS: TestAccAWSAPIGatewayMethodResponse_basic (119.79s)
=== RUN   TestAccAWSAPIGatewayMethod_basic
--- PASS: TestAccAWSAPIGatewayMethod_basic (59.41s)
=== RUN   TestAccAWSAPIGatewayMethod_customauthorizer
--- PASS: TestAccAWSAPIGatewayMethod_customauthorizer (79.44s)
=== RUN   TestAccAWSAPIGatewayModel_basic
--- PASS: TestAccAWSAPIGatewayModel_basic (39.39s)
=== RUN   TestAccAWSAPIGatewayResource_basic
--- PASS: TestAccAWSAPIGatewayResource_basic (35.46s)
=== RUN   TestAccAWSAPIGatewayRestApi_basic
--- PASS: TestAccAWSAPIGatewayRestApi_basic (53.67s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/aws	1115.501s
```

**Comment by kwilczynski** [comment]
I need to verify acceptance test for S3 bucket, and it should be good to go.

**Comment by kwilczynski** [comment]
The S3 Bucket acceptance test is passing:

```
$ make testacc TEST=./builtin/providers/aws TESTARGS='-run=TestAccAWSS3Bucket_'
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2016/10/04 18:30:02 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/aws -v -run=TestAccAWSS3Bucket_ -timeout 120m
=== RUN   TestAccAWSS3Bucket_importBasic
--- PASS: TestAccAWSS3Bucket_importBasic (57.73s)
=== RUN   TestAccAWSS3Bucket_importWithPolicy
--- PASS: TestAccAWSS3Bucket_importWithPolicy (60.72s)
=== RUN   TestAccAWSS3Bucket_Notification
--- PASS: TestAccAWSS3Bucket_Notification (153.89s)
=== RUN   TestAccAWSS3Bucket_NotificationWithoutFilter
--- PASS: TestAccAWSS3Bucket_NotificationWithoutFilter (63.79s)
=== RUN   TestAccAWSS3Bucket_basic
--- PASS: TestAccAWSS3Bucket_basic (46.68s)
=== RUN   TestAccAWSS3Bucket_acceleration
--- PASS: TestAccAWSS3Bucket_acceleration (95.81s)
=== RUN   TestAccAWSS3Bucket_RequestPayer
--- PASS: TestAccAWSS3Bucket_RequestPayer (91.07s)
=== RUN   TestAccAWSS3Bucket_Policy
--- PASS: TestAccAWSS3Bucket_Policy (131.62s)
=== RUN   TestAccAWSS3Bucket_UpdateAcl
--- PASS: TestAccAWSS3Bucket_UpdateAcl (94.10s)
=== RUN   TestAccAWSS3Bucket_Website_Simple
--- PASS: TestAccAWSS3Bucket_Website_Simple (153.28s)
=== RUN   TestAccAWSS3Bucket_WebsiteRedirect
--- PASS: TestAccAWSS3Bucket_WebsiteRedirect (151.93s)
=== RUN   TestAccAWSS3Bucket_WebsiteRoutingRules
--- PASS: TestAccAWSS3Bucket_WebsiteRoutingRules (108.90s)
=== RUN   TestAccAWSS3Bucket_shouldFailNotFound
--- PASS: TestAccAWSS3Bucket_shouldFailNotFound (30.07s)
=== RUN   TestAccAWSS3Bucket_Versioning
--- PASS: TestAccAWSS3Bucket_Versioning (136.25s)
=== RUN   TestAccAWSS3Bucket_Cors
--- PASS: TestAccAWSS3Bucket_Cors (106.47s)
=== RUN   TestAccAWSS3Bucket_Logging
--- PASS: TestAccAWSS3Bucket_Logging (94.12s)
=== RUN   TestAccAWSS3Bucket_Lifecycle
--- PASS: TestAccAWSS3Bucket_Lifecycle (151.96s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/aws	1728.437s
```

**Comment by kwilczynski** [comment]
Related to #9213.

File: builtin/providers/aws/validators.go [hunk 1]
```diff
@@ -467,3 +482,18 @@ func validateJsonString(v interface{}, k string) (ws []string, errors []error) {
 	}
 	return
 }
+
+func sentenceJoin(conjunction string, m map[string]bool) string {
```
Comment (stack72) [inline builtin/providers/aws/validators.go]: is this a validator?

File: builtin/providers/aws/validators.go [hunk 2]
```diff
@@ -343,9 +345,22 @@ func validateCIDRNetworkAddress(v interface{}, k string) (ws []string, errors []
...
+		"OPTIONS": true,
+		"PATCH":   true,
+		"POST":    true,
+		"PUT":     true,
+	}
+
+	if _, ok := validMethods[value]; !ok {
 		errors = append(errors, fmt.Errorf(
-			"%q must be one of 'GET', 'HEAD', 'OPTIONS', 'PUT', 'POST', 'PATCH', 'DELETE'", k))
+			"%q contains an invalid method %q. Valid methods are either %s",
```
Comment (stack72) [inline builtin/providers/aws/validators.go]: Does the %s here work for the map[string]bool?

File: builtin/providers/aws/resource_aws_s3_bucket_object.go [hunk 3]
```diff
@@ -293,25 +292,10 @@ func validateS3BucketObjectAclType(v interface{}, k string) (ws []string, errors
...
-
-		words[length-1] = fmt.Sprintf("or %s", words[length-1])
-		return strings.Join(words, ", ")
-	}
-
 	if _, ok := cannedAcls[value]; !ok {
 		errors = append(errors, fmt.Errorf(
 			"%q contains an invalid canned ACL type %q. Valid types are either %s",
-			k, value, sentenceJoin(cannedAcls)))
+			k, value, sentenceJoin("or", cannedAcls)))
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_s3_bucket_object.go]: I'm tempted to say we shouldn't touch this as part of this PR. It isn't related at all to the API Gateway validator. I think this should be a separate PR IMO

File: builtin/providers/aws/validators.go (hunk 2 above)
Comment (kwilczynski) [inline builtin/providers/aws/validators.go]: The helper outputs a string with items surrounded by quotes, so it should be fine.

File: builtin/providers/aws/resource_aws_s3_bucket_object.go (hunk 3 above)
Comment (kwilczynski) [inline builtin/providers/aws/resource_aws_s3_bucket_object.go]: Ah, this snook up by accident. I was rebasing recently.

**Comment by kwilczynski** [comment]
@stack72 amended :) I removed the extra helper, and corrected the S3 change.
